ui/apps/pmm/src/themes/setTheme.ts

```diff
@@ -0,0 +1,124 @@
+import { useContext, useRef } from 'react';
+import { ColorModeContext } from '@percona/design';
+import messenger from 'lib/messenger';
+import { grafanaApi } from 'api/api';
+
+type Mode = 'light' | 'dark';
+
+/** Normalizes any incoming value to 'light' | 'dark'. */
+function normalizeMode(v: unknown): Mode {
+  if (typeof v === 'string' && v.toLowerCase() === 'dark') return 'dark';
+  if (v === true) return 'dark';
+  return 'light';
+}
+
+/** Idempotently applies theme attributes to <html>. */
+function applyDocumentTheme(mode: Mode) {
+  const html = document.documentElement as HTMLElement & {
+    style: CSSStyleDeclaration & { colorScheme?: string };
+  };
+  const scheme = mode === 'dark' ? 'percona-dark' : 'percona-light';
+  const wantDark = mode === 'dark';
+  const hasDarkClass = html.classList.contains('dark');
+
+  // If everything (including the Tailwind 'dark' class) is already correct, skip.
+  if (
+    html.getAttribute('data-theme') === mode &&
+    html.getAttribute('data-md-color-scheme') === scheme &&
+    html.style.colorScheme === mode &&
+    hasDarkClass === wantDark
+  ) {
+    return;
+  }
+
+  html.classList.toggle('dark', wantDark);
+
+  html.setAttribute('data-theme', mode);
+  html.setAttribute('data-md-color-scheme', scheme);
+  html.style.colorScheme = mode;
+}
+
+/**
+ * useSetTheme centralizes theme changes for:
+ * - local PMM UI (left)
+ * - persistence to Grafana preferences
+ * - broadcasting to Grafana iframe (right)
+ */
+export function useSetTheme() {
```

Comment:
It would be good to centralize the theme handling/changes into a separate provider.

Should load the current theme based on the localStorage/userPreferences and store changes in user preferences when theme changes.